R/http_methods.R

```diff
@@ -0,0 +1,185 @@
+#' fetch_version()
+#' # fetch the Chromium/Chrome protocol
+#' protocol <- fetch_protocol()
+#'
+#' # get the list of opened tabs
+#' list_targets()
+#' # open an url in a new tab
+#' new_target <- new_tab(url = "http://r-project.org")
+#' # the new tab is referenced in the list of opened tabs
+#' list_targets()
+#' # inspect a target in a web browser
+#' if(interactive()) inspect_target(target_id = new_target$id)
+#' # close the previous created tab
+#' close_target(target_id = new_target$id)
+#'
+#' chrome$close()
+#' }
+NULL
+
+#' @rdname http-methods
+#' @export
+fetch_version <- function(host = "localhost", port = 9222, secure = FALSE) {
+  fetch_json(host, port, secure, "version")
+}
+
+#' @rdname http-methods
+#' @export
+list_targets <- function(host = "localhost", port = 9222, secure = FALSE) {
+  fetch_json(host, port, secure, "list")
+}
+
+#' @rdname http-methods
+#' @export
+fetch_protocol <- function(host = "localhost", port = 9222, secure = FALSE, local = FALSE) {
+  # if the local protocol is fetched, return early
+  if(isTRUE(local)) {
+    return(read_local_protocol())
+  }
+  fetch_json(host, port, secure, "protocol")
+}
+
+#' @rdname http-methods
+#' @export
+new_tab <- function(host = "localhost", port = 9222, secure = FALSE, url = NULL) {
+  if(!is.null(url)) {
+    assert_that(is_scalar_character(url))
+  }
+  fetch_json(host, port, secure, "new", url)
+}
+
+#' @rdname http-methods
+#' @export
+activate_target <- function(
+  host = "localhost", port = 9222, secure = FALSE, target_id
+) {
+  target_method(host, port, secure, target_id, "activate")
+}
+
+#' @rdname http-methods
+#' @export
+close_target <- function(
+  host = "localhost", port = 9222, secure = FALSE, target_id
+) {
+  target_method(host, port, secure, target_id, "close")
+}
+
+#' @rdname http-methods
+#' @export
+inspect_target <- function(
+  host = "localhost", port = 9222, secure = FALSE, target_id
+) {
+  assert_that(is_scalar_character(target_id))
+  targets <- list_targets(host, port, secure)
+  ids <- purrr::map_chr(targets, "id")
+  index <- purrr::detect_index(ids, ~ identical(.x, target_id))
+  if(index == 0) {
+    stop("Wrong target_id.")
+  }
+  if(!interactive()) {
+    warning("The inspect method can only be used in an interactive session.")
+    return(NULL)
+  }
+  path <- purrr::pluck(targets, index, "devtoolsFrontendUrl")
+  url <- build_http_url(host, port, secure, path)
+  browse_url(url)
+}
+
+fetch_json <- function(host, port, secure, method, query = NULL) {
+  check_host_port_args(host, port)
+  url <- build_http_url(host, port, secure, path = c("json", method), query)
+  from_json(url)
+}
+
+target_method <- function(host, port, secure, target_id, method) {
+  check_host_port_args(host, port)
+  assert_that(is_scalar_character(target_id))
+  url <- build_http_url(host, port, secure, path = c("json", method, target_id))
+  res <- httr::GET(url, httr::use_proxy(""))
+  !httr::http_error(res)
```

Comment:
The problem I see with throwing errors too early is that in many cases we have to return a rejected promise instead of throwing an error. So, we should have to use `tryCatch(...)` when calling them in an async function.

If we want to throw errors, I think we should implement an `async` option and return a promise if `async` is `TRUE`.

FMPOV, in this case, the use of a HTTP error is a pure convention. I do not consider this case as an exception (the remote host has sent a HTTP response). 
If the remote could not be reached we would get an exception/error. 
I see this HTTP error only as a mean to tell that the target ID was wrong. Do we need to treat this case as an exception? I have serious doubt for now. But if you think that it is required, I have no problem with that.

Reply:
As this will not be used because internal, it is not so important. It feels just strange to have FALSE when you do something that not suppose to work at all. If we had a real use of `activateTab` to deal with several tabs, this would be more important, but I guess currently FALSE is more than ok.